confidential-datahub/docs/SEALED_SECRET.md

```diff
@@ -0,0 +1,152 @@
+# Kubernetes Sealed Secret
+
+## Introduction
+
+CoCo’s [threat model](https://github.com/confidential-containers/confidential-containers/blob/main/trust_model_personas.md)
+excludes the Kubernetes Control Plane and Host components from the
+Trusted Compute Base (TCB). With this stricter threat model, CoCo
+runtimes cannot take advantage of the existing infrastructure. to
+protect [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/),
+this document introduces Sealed Secrets that supports in-pod decryption.
+
+A Sealed Secret is a set of metadata to get the plaintext of the secret.
+A Sealed Secret will be unsealed inside the TEE pod transparently to users,
+and the unsealing occurs only if the remote attestation process passes,
+which means the TEE environment is as expected. Also, Sealed Secret can
+leverage commercial KMS/Secret Manager(Vault) productions in the unsealing
+process.
+
+## Format
+
+Due to different sealing types, there are two different formats of Sealed
+Secrets.
+
+### KMS
+
+This kind of secret uses envelope encryption scheme. A wrapping key is used
+to encrypt the plaintext secret value. A sealing key insde a KMS is used to
+seal the wrapping key. That is
+$```
+Sealed Secret := \{Enc_{kms key}(Wrapping Key), Enc_{Wrapping Key}(secret value)\}
+```$
+
+The format of the KMS type Sealed Secret is
+```json
+{
+	"version": "0.1.0",
+	"type": "kms",
+	"provider": "xxx",
+	"key_id": "xxx",
+	"encrypted_key": "ab27dc=", 
+	"encrypted_data": "xxx",
+	"wrap_type": "A256GCM",
+	"iv": "xxx",
+	"annotations": {
+		"region_id": "cn-hangzhou",
+		"endpoint": "xxx",
+		"signature_method": "xxx"
+	}
+}
+```
+Here,
+- `version`: indicates the format version of the Sealed Secret. Currently `0.1.0`.
+- `type`: MUST be `kms`, indicating this is a KMS type Sealed Secret
+- `provider`: indicates the provider of the __kms key__. This field determines
+how to use the `annotations` field and `key_id` field to decrypt the `encrypted_key`
+- `key_id`: To uniquely distinguish the __kms key__ used to encrypt the __wrapping key__,
+which is always used by the provider plugin.
+- `encrypted_key`: Encrypted __wrapping key__ by the `provider`. Base64 encoded.
+- `encrypted_data`: Encrypted __secret value__ by the `encrypted_key`. Base64 encoded.
+- `wrap_type`: The algorithm used by __wrapping key__ to encrypt the __secret value__.
+`A256GCM` (AES256-GCM) preferred.
+- `iv`: The Initial Vector used in the process of __wrapping key__ encrypting __secret value__.
+Base64 encoded.
+- `annotations`: A key-value Map. KMS specific information used by the KMS driver to
+decrypt `encrypted_key` into a plaintext of __wrapping key__.
+
+### Vault
+
+A Vault secret leverages secret manager mechanism. Vault secret does not require any
+encryption. It only contains metadata with which the plaintext of the __secret value__
+can be retrieved. The format is
```

Comment:
Oh, yes. From a programmer's point of view, the `type` field tells it is a vault secret.
Then `provider` field tells which driver to use.
Finally `name` shows the location and it is same as `id` semantically. `annotations` are also used to locate the vault secret if `name` is not enough. The `name` and `annotations` are all inputs to the driver to get the secret.

Reply:
Ok, I am still a bit skeptical, but I guess we can see when we try to actually implement. This seems like enough information to request a secret from a KMS, but how do we setup the connection to the KMS in the first place? is there other configuration information provided elsewhere?

Reply:
Yes. To init the connection to a KMS/Vault server, there should be two types of parameters:
- public parameters: like the region id of the KMS, the api address of KMS, the signature method of the request, etc. These parameters should be placed inside `Annotation` field
- private parameters: like the access key to the KMS. These parameters should be include the `new()` function of the KMS driver. Read from mem-filesystem recommended, like `/run/xxx`. The architecture of CDH has a `auth layer`, this can integrate some mechanisms to fetch the access keys. Initially, we can implement one that uses KBS to get the access keys.

Reply:
Ok, I think there may be some questions of scope i.e. should the secret be able to point to different instances of a vault that have the same type. But we can worry about that kind of thing later.